igor-web/src/test/groovy/com/netflix/spinnaker/igor/wercker/WerckerBuildMonitorSpec.groovy

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) 2017, 2018, Oracle Corporation and/or its affiliates. All rights reserved.
+ *
+ * The contents of this file are subject to the Apache License Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * If a copy of the Apache License Version 2.0 was not distributed with this file,
+ * You can obtain one at https://www.apache.org/licenses/LICENSE-2.0.html
+ */
+package com.netflix.spinnaker.igor.wercker
+
+import com.netflix.spectator.api.NoopRegistry
+import com.netflix.spinnaker.igor.IgorConfigurationProperties
+import com.netflix.spinnaker.igor.config.WerckerProperties
+import com.netflix.spinnaker.igor.config.WerckerProperties.WerckerHost
+import com.netflix.spinnaker.igor.history.EchoService
+import com.netflix.spinnaker.igor.model.BuildServiceProvider
+import com.netflix.spinnaker.igor.service.BuildMasters
+import com.netflix.spinnaker.igor.wercker.model.Application
+import com.netflix.spinnaker.igor.wercker.model.Owner
+import com.netflix.spinnaker.igor.wercker.model.Pipeline
+import com.netflix.spinnaker.igor.wercker.model.Run
+import com.netflix.spinnaker.kork.eureka.RemoteStatusChangedEvent
+
+import java.util.concurrent.TimeUnit
+
+import rx.schedulers.TestScheduler
+import spock.lang.Specification
+
+class WerckerBuildMonitorSpec extends Specification {
+
+    WerckerCache cache = Mock(WerckerCache)
+    WerckerBuildMonitor monitor
+    EchoService echoService = Mock(EchoService)
+    WerckerClient client
+    WerckerService mockService = Mock(WerckerService)
+    WerckerService werckerService
+    String werckerDev = 'https://dev.wercker.com/'
+    String master = 'WerckerTestMaster'
+
+    void setup() {
+        client = Mock(WerckerClient)
+        werckerService = new WerckerService(
+                new WerckerHost(name: master, address: werckerDev), cache, client)
+    }
+
+    final MASTER = 'MASTER'
+    final pipeline = 'myOrg/myApp/pollingTest'
+    final List<String> APPS = []
+    final TestScheduler scheduler = new TestScheduler()
+
+    void 'no finished run'() {
+        given:
+        cache.getJobNames(MASTER) >> ['pipeline']
```

Comment:
Most of this setup code is duplicated with the next test (and appears on a number of them).  It would probably make sense to pull this into a function, or just have multiple tests under the same `given`.  I think all the tests (except maybe the last one with multiple pipelines) could just have one setup and a data table (or multiple when/then).

Reply:
Moved the duplicated setup to mockBuildMasters().